### PR TITLE
opt: fully normalize partial index predicates

### LINKED ISSUE #54819: sql: failure to create partial index: "index <name> is a partial index that does not contain all the rows needed to execute this query"
**Describe the problem**

I'm not sure if this is a bug or an unclear message describing a limitation of CockroachDB.

**To Reproduce**
```sql
CREATE TABLE public.indexes_article (
    id INT8 NOT NULL DEFAULT unique_rowid(),
    headline VARCHAR(100) NOT NULL,
    pub_date TIMESTAMPTZ NOT NULL,
    published BOOL NOT NULL,
    CONSTRAINT "primary" PRIMARY KEY (id ASC),
    INDEX indexes_article_headline_pub_date_b992dbba_idx (headline ASC, pub_date ASC),
    FAMILY "primary" (id, headline, pub_date, published)
)

CREATE INDEX "recent_article_idx" ON "indexes_article" ("pub_date") WHERE "pub_date" IS NOT NULL;

ERROR: verify-idx-count: index "recent_article_idx" is a partial index that does not contain all the rows needed to execute this query
```
**Expected behavior**

No error, or if this is a limitation of CockroachDB, perhaps the error message could be more descript?

**Environment:**
 - CockroachDB version v20.2.0-alpha.1-3636-g3c81612122 @ 2020/09/24 20:37:23

**Additional context**

This is from a Django test that works on SQLite and PostgreSQL.

### PR DESCRIPTION
**Pull request by rytaft**
Prior to this commit, only a select few normalization rules were applied
to partial index predicates. This commit ensures that all applicable
normalization rules are applied by building a temporary `Select` expression
with the index predicates. This is necessary to keep partial index predicates
in sync with predicates that may be used as part of a query, so that
implication can be correctly proven.

Fixes #54819

Release note (bug fix): Fixed an error that could occur during planning
when attpempting to create or use a partial index. The error implied that
the partial index could not be used (e.g., "my_index is a partial index
that does not contain all the rows needed to execute this query") when in
fact the partial index could be used correctly.

### CODE REVIEW

**Comment by rytaft** [comment]
___
*[pkg/sql/opt/optbuilder/partial_index.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55057#-MIUblLK0lnCNFAFlL86:-MIUblLK0lnCNFAFlL87:bb98ik) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ff98898416c73f0bf874fd0bb9a7392490c567/pkg/sql/opt/optbuilder/partial_index.go#L73)):*
> ```Go
> // because different filters should be treated separately. For example,
> // x < 0 AND x > 0 is not a contradiction when used as an arbiter predicate.
> func (b *Builder) buildArbiterPredicate(tableScope *scope, expr tree.Expr) memo.FiltersExpr {
> ```

I don't fully understand the rules about arbiter predicates, so please check this carefully. Note that I completely separated the code paths, because it seems like the rules around non-immutable operators need not apply to arbiter predicates (and using the error message "partial index predicate is not immutable" to refer to an arbiter predicate seems potentially confusing).

**Review by mgartner — APPROVED** [review]
This is fantastic! This should put this type of issue to rest once and for all. :tada: 

I left some comments about arbiters, but they are not fully baked. I'm sure @raduberinde has much wiser thoughts on the matter.

> Reviewed 12 of 14 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55057#-:-MIV2Qng9Uxl_vXLV5Ay:b-wtr3dv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/optbuilder/partial_index.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55057#-MIUblLK0lnCNFAFlL86:-MIUs8kQ6VOUln_RzC-N:bq4bega) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ff98898416c73f0bf874fd0bb9a7392490c567/pkg/sql/opt/optbuilder/partial_index.go#L73)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I don't fully understand the rules about arbiter predicates, so please check this carefully. Note that I completely separated the code paths, because it seems like the rules around non-immutable operators need not apply to arbiter predicates (and using the error message "partial index predicate is not immutable" to refer to an arbiter predicate seems potentially confusing).
</blockquote></details>

I agree that the error message "partial index predicate is not immutable" would be confusing. I'm trying to think of an example where a non-immutable arbiter predicate could imply an immutable partial index predicate.

Just thinking out loud for a minute—if a stable operator in the arbiter predicate is folded into an immutable expression, then it could imply an immutable partial index predicate. That seems dangerous. From a quick test on Postgres, it prevents this from happening, though I can't be sure whether its because stable operator folding is disable or some other reason (Note that I'm using the stable `current_user` builtin because it was easier for my vacation brain to reason about than timezone-related functions or any other stable functions):

```
marcus=# create table t (t text);
CREATE TABLE

marcus=# create unique index i on t (t) where t > 'marcus';
CREATE INDEX

marcus=# insert into t values ('z') on conflict (t) where t > current_user do nothing;
ERROR:  there is no unique or exclusion constraint matching the ON CONFLICT specification
```

I'm not sure, but my guess is that if you added a test like this to the PR, the insert would succeed. So, I think we should at least disable stable folds in `buildArbiterPredicate`, like we do in `buildPartialIndexPredicate`.

Also, I'd lean towards disallowing non-immutable expressions for arbiter predicates entirely, because it seems like the safest thing to do. In addition, if it is impossible for a non-immutable expression to imply an immutable expression (I can't think of a counter-example), then behavior of erring here while building the predicate will be the same as erring when failing to select the arbiter index, from the users's perspective.

Lastly, I _think_ the code paths (besides the error message) could be identical. The contradiction case in the function comment, `x < 0 AND x > 0` should normalize to `false`. `false` implies any expression because `false` denotes an empty set of results and all sets contain the empty set. The implicator was recently fixed to support this in #54718. So with full normalization we'd match Postgres behavior in the given example:

```
marcus=# create table t (a int);
CREATE TABLE

marcus=# create unique index i on t (a) where a > 0;
CREATE INDEX

marcus=# insert into t values (1) on conflict (a) where a > 0 and a < 0 do nothing;
INSERT 0 1
```

But, oddly, Postgres doesn't allow _any_ contradiction to be used to select an arbiter. I'm not exactly sure why. Maybe this is a bug? Or missing feature?

```
marcus=# insert into t values (1) on conflict (a) where false do nothing;
ERROR:  there is no unique or exclusion constraint matching the ON CONFLICT specification

marcus=# insert into t values (1) on conflict (a) where a < -10 and a = -5 do nothing;
ERROR:  there is no unique or exclusion constraint matching the ON CONFLICT specification
```

So TLDR, I think we should:

1. **Definitely** disable stable folding in `buildArbiterPredicate`.
2. **Probably** error when building  a non-immutable arbiter predicate with the same error message as not finding an arbiter: "there is no unique or exclusion constraint matching the ON CONFLICT specification"
3. **Maybe** fully normalization of the arbiter.

**Review by rytaft — COMMENTED** [review]
TFTR! Updated.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55057#-:-MIVTbmE-tldUmKVYfdq:b-p0fj95)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @RaduBerinde)

___
*[pkg/sql/opt/optbuilder/partial_index.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55057#-MIUblLK0lnCNFAFlL86:-MIVTgXq41Wz5hPAEll9:b71t2nj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ff98898416c73f0bf874fd0bb9a7392490c567/pkg/sql/opt/optbuilder/partial_index.go#L73)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I agree that the error message "partial index predicate is not immutable" would be confusing. I'm trying to think of an example where a non-immutable arbiter predicate could imply an immutable partial index predicate.

Just thinking out loud for a minute—if a stable operator in the arbiter predicate is folded into an immutable expression, then it could imply an immutable partial index predicate. That seems dangerous. From a quick test on Postgres, it prevents this from happening, though I can't be sure whether its because stable operator folding is disable or some other reason (Note that I'm using the stable `current_user` builtin because it was easier for my vacation brain to reason about than timezone-related functions or any other stable functions):

```
marcus=# create table t (t text);
CREATE TABLE

marcus=# create unique index i on t (t) where t > 'marcus';
CREATE INDEX

marcus=# insert into t values ('z') on conflict (t) where t > current_user do nothing;
ERROR:  there is no unique or exclusion constraint matching the ON CONFLICT specification
```

I'm not sure, but my guess is that if you added a test like this to the PR, the insert would succeed. So, I think we should at least disable stable folds in `buildArbiterPredicate`, like we do in `buildPartialIndexPredicate`.

Also, I'd lean towards disallowing non-immutable expressions for arbiter predicates entirely, because it seems like the safest thing to do. In addition, if it is impossible for a non-immutable expression to imply an immutable expression (I can't think of a counter-example), then behavior of erring here while building the predicate will be the same as erring when failing to select the arbiter index, from the users's perspective.

Lastly, I _think_ the code paths (besides the error message) could be identical. The contradiction case in the function comment, `x < 0 AND x > 0` should normalize to `false`. `false` implies any expression because `false` denotes an empty set of results and all sets contain the empty set. The implicator was recently fixed to support this in #54718. So with full normalization we'd match Postgres behavior in the given example:

```
marcus=# create table t (a int);
CREATE TABLE

marcus=# create unique index i on t (a) where a > 0;
CREATE INDEX

marcus=# insert into t values (1) on conflict (a) where a > 0 and a < 0 do nothing;
INSERT 0 1
```

But, oddly, Postgres doesn't allow _any_ contradiction to be used to select an arbiter. I'm not exactly sure why. Maybe this is a bug? Or missing feature?

```
marcus=# insert into t values (1) on conflict (a) where false do nothing;
ERROR:  there is no unique or exclusion constraint matching the ON CONFLICT specification

marcus=# insert into t values (1) on conflict (a) where a < -10 and a = -5 do nothing;
ERROR:  there is no unique or exclusion constraint matching the ON CONFLICT specification
```

So TLDR, I think we should:

1. **Definitely** disable stable folding in `buildArbiterPredicate`.
2. **Probably** error when building  a non-immutable arbiter predicate with the same error message as not finding an arbiter: "there is no unique or exclusion constraint matching the ON CONFLICT specification"
3. **Maybe** fully normalization of the arbiter.
</blockquote></details>

Nice -- I was able to combine the logic for the two. I was just confused before when I thought they needed to be separate. I've also fixed this so it's returning the correct error now for arbiters with non-immutable operators.

**Comment by rytaft** [comment]
@raduberinde do you agree we should backport this? If so, ok to merge? (Want to make sure I get it in before today's deadline)

**Review by RaduBerinde — APPROVED** [review]
:lgtm: Yes, I agree we should backport.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55057#-:-MIVp0KJBhh1BSSQo_5O:bcbbq2c)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/logictest/testdata/logic_test/partial_index, line 897 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55057#-MIVo3ra8BPw1RvIt_lZ:-MIVo3rbDNzC4MB01Vwj:bvqda6f) ([raw file](https://github.com/cockroachdb/cockroach/blob/015790a85b98fac87e97f27eb54dd3d971cb8ce0/pkg/sql/logictest/testdata/logic_test/partial_index#L897)):*
> ```
> # It is not valid to use non-immutable functions in an arbiter predicate with a
> # partial index as the arbiter index.
> statement error pgcode 42P10 there is no unique or exclusion constraint matching the ON CONFLICT specification
> ```

Hm, this would error out the same regardless if we folded stable functions or not.. Maybe an example where it does actually fold to the index predicate is better (e.g. `b < CASE WHEN now() > '1980-01-01' THEN 0 ELSE 100 END`). Though I'm ambivalent as to whether we should error out in this case or not.
___
*[pkg/sql/opt/memo/expr_format.go, line 797 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55057#-MIVoupKAiW7BGMguI1k:-MIVoupKAiW7BGMguI1l:b-86mog) ([raw file](https://github.com/cockroachdb/cockroach/blob/015790a85b98fac87e97f27eb54dd3d971cb8ce0/pkg/sql/opt/memo/expr_format.go#L797)):*
> ```Go
> 		// Show empty Filters expression as "filters (true)".
> 		if scalar.ChildCount() == 0 {
> 			f.Buffer.WriteString("filters (true)")
> ```

Does this fix something?

**Review by rytaft — COMMENTED** [review]
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55057#-:-MIVtMKS8XcCOlfz4-nZ:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @RaduBerinde)

___
*[pkg/sql/logictest/testdata/logic_test/partial_index, line 897 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55057#-MIVo3ra8BPw1RvIt_lZ:-MIVsD3S4H91NtiY1CmM:b-vmcqo5) ([raw file](https://github.com/cockroachdb/cockroach/blob/015790a85b98fac87e97f27eb54dd3d971cb8ce0/pkg/sql/logictest/testdata/logic_test/partial_index#L897)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Hm, this would error out the same regardless if we folded stable functions or not.. Maybe an example where it does actually fold to the index predicate is better (e.g. `b < CASE WHEN now() > '1980-01-01' THEN 0 ELSE 100 END`). Though I'm ambivalent as to whether we should error out in this case or not.
</blockquote></details>

Good idea - done. Error-ing out in this case matches Postgres' behavior.
___
*[pkg/sql/opt/memo/expr_format.go, line 797 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55057#-MIVoupKAiW7BGMguI1k:-MIVsKMPEb1GRHkCMeRb:b-4mv40c) ([raw file](https://github.com/cockroachdb/cockroach/blob/015790a85b98fac87e97f27eb54dd3d971cb8ce0/pkg/sql/opt/memo/expr_format.go#L797)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Does this fix something?
</blockquote></details>

Yea, the partial indexes were just printing out as "filters (true)" instead of including the partial index name as a label.

**Review by RaduBerinde — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55057#-:-MIVvJ5mCYHfMPV3ntjU:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)
